src/libfetchers/git.cc

```diff
@@ -24,6 +26,110 @@ static std::string readHead(const Path & path)
+  tree,
+  commit
+};
+
+string getBlob(const Path & path, const string & gitrev, const string & blobhash) {
+  return runProgram("git", true, { "-C", path, "cat-file", "-p", blobhash });
+}
+
+static std::map<string, std::pair<gitType, string>> gitTreeList(const Path & path, const string & treehash) {
+  auto treeOutput = runProgram("git", true, { "-C", path, "cat-file", "-p", treehash });
+  auto lines = tokenizeString<std::vector<string>>(treeOutput, "\n");
+  std::map<string, std::pair<gitType, string>> results;
+  for (auto line : lines) {
+    auto words = tokenizeString<std::vector<string>>(line);
+    auto hash = words[2];
+    auto name = words[3];
+    gitType type;
+    if (words[1] == "blob") type = blob;
+    else if (words[1] == "tree") type = tree;
+    else if (words[1] == "commit") type = commit;
+    else assert(0); // FIXME
+    results.insert(std::pair{name, std::pair{type, hash}});
+  }
+
+  return results;
+}
+
+std::map<string,string> parseSubmodules(const string & contents) {
+  auto lines = tokenizeString<std::vector<string>>(contents, "\n");
+  std::map<string,string> results;
+  string path;
+  for (auto line : lines) {
+    auto words = tokenizeString<std::vector<string>>(line);
+    if (words[1] != "=") continue;
+
+    if (words[0] == "path") path = words[2];
+    if (words[0] == "url") {
+      string url = words[2];
+      results.insert(std::pair{path, url});
+    }
+  }
+  return results;
+}
+
+string findCommitHash(const Path & path, const std::map<string, std::pair<gitType, string>> & _entries, const string & pathToFind) {
+  auto entries = _entries;
+
+  auto tokens = tokenizeString<std::vector<string>>(pathToFind, "/");
+  std::pair<gitType, string> x;
+  for (auto token : tokens) {
+    auto i = entries.find(token);
+    if (i == entries.end()) return "fail";
+
+    x = i->second;
+    if (x.first == tree) entries = gitTreeList(path, x.second);
+  }
+  if (x.first != commit) return "fail";
+
+  return x.second;
+}
+
+static Attrs readSubmodules(const Path & path, const string & gitrev)
+{
+    auto rootTree = getRootTree(path, gitrev);
+    printTalkative("root tree is %s", rootTree);
+    auto entries = gitTreeList(path, rootTree);
+    Attrs attrs;
+
+    auto i = entries.find(".gitmodules");
+
+    if (i == entries.end()) return attrs;
+
+    auto submodules = getBlob(path, gitrev, i->second.second);
+
+    printTalkative("submodule file %s", submodules);
+
+    auto parsedModules = parseSubmodules(submodules);
+    for (auto & [subPath, url] : parsedModules) {
+      printTalkative("submodule %s fetched from %s", subPath, url);
+      auto commitHash = findCommitHash(path, entries, subPath);
+      printTalkative("found %s", commitHash);
+
+      static const std::regex barePathRegex("^/.*$");
+      std::string prefix = "git+";
+      if (std::regex_match(url, barePathRegex))
+          prefix = prefix + "file://";
+
+
+      attrs.emplace(subPath, prefix + url + "?allRefs=1&rev=" + commitHash);
```

Comment:
If the submodule commit is available only in the local clone and not in the remote, we need to pass the former + submodule path as `url` here I think.

Reply:
Drop "only" from my sentence to avoid redundant fetches as well